File: src/main/scala/com/mikea/gae/rx/RxImplConfigGen.java
package com.mikea.gae.rx;

import com.google.common.reflect.TypeToken;
import com.google.inject.Inject;
import com.google.inject.Injector;
import com.mikea.gae.rx.base.IDisposable;
import com.mikea.gae.rx.base.IObserver;

import java.io.IOException;
import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

class RxImplConfigGen implements Rx {
    private final Injector injector;
    private Set<String> cronSpecifications = new HashSet<>();
    private Set<String> taskQueues = new HashSet<>();

    @Inject
    RxImplConfigGen(Injector injector) {
        this.injector = injector;
    }

    @Override
    public RxStream<RxCronEvent> cron(String specification) {
        cronSpecifications.add(specification);
        return new RxConfigGenStream<>(this);
    }

    @Override
    public Injector getInjector() {
        return injector;
    }

    @Override
    public RxStream<RxUploadEvent> uploads() {
        return new RxConfigGenStream<>(this);
    }

    @Override
    public <T extends Serializable> IObserver<RxTask<T>> taskqueue(String queueName) {
        taskQueues.add(queueName);

        return new IObserver<RxTask<T>>() {
            @Override
            public void onCompleted() {
                throw new UnsupportedOperationException();
            }

            @Override
            public void onError(Exception e) {
                throw new UnsupportedOperationException();
            }

            @Override
            public void onNext(RxTask<T> value) throws IOException {
                throw new UnsupportedOperationException();
            }
        };
    }

    @Override
    public <T extends Serializable> RxStream<RxTask<T>> tasks(String uploads, Class<T> payloadClass) {
        return new RxConfigGenStream<>(this);
    }

    @Override
    public <T extends Serializable> RxStream<RxTask<T>> tasks(String queueName, TypeToken<T> typeToken) {
        return new RxConfigGenStream<>(this);
    }

    @Override
    public RxStream<RxVersionUpdateEvent> updates() {
        return new RxConfigGenStream<>(this);
    }

    @Override
    public RxStream<RxInitializationEvent> initialized() {
        return new RxConfigGenStream<>(this);
    }

    public void generateConfigs() {
        System.out.println("----- cron.xml -----");
        System.out.println(
                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                        "<cronentries>");
        for (String cronSpecification : cronSpecifications) {
            System.out.println(
                    "    <cron>\n" +
                            "        <url>" + RxImpl.getCronUrl(cronSpecification) + "</url>\n" +
                            "        <schedule>" + cronSpecification + "</schedule>\n" +
                            "    </cron>");
        }
        System.out.println("</cronentries>\n");
        System.out.println("----- queue.xml -----");
        System.out.println(
                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                        "<queue-entries>");
        for (String queue : taskQueues) {
            System.out.println(
                    "    <queue>\n" +
                            "        <name>" + queue + "</name>\n" +
                            "    </queue>");
        }
        System.out.println("</queue-entries>\n");
    }

    private static class RxConfigGenStream<T> extends RxStream<T> {
        protected RxConfigGenStream(Rx rx) {
            super(rx);
        }

        @Override
        public IDisposable subscribe(IObserver<T> observer) {
            return new IDisposable() {
                @Override
                public void dispose() {
                }
            };
        }
    }
}

File: src/main/scala/com/mikea/gae/rx/RxImplConfigGen.scala
package com.mikea.gae.rx

import com.google.common.reflect.TypeToken
import com.google.inject.Inject
import com.google.inject.Injector
import com.mikea.gae.rx.base.{IObservable, IObserver}
import java.io.Serializable
import com.mikea.gae.rx.RxImplConfigGen.RxConfigGenStream
import scala.collection.immutable.HashSet

object RxImplConfigGen {
  class RxConfigGenStream[T] (_injector: Injector) extends IObservable[T] {
    def subscribe(observer: IObserver[T]) = {
      throw new UnsupportedOperationException()
    }

    def instantiate[C](aClass: Class[C]) = _injector.getInstance(aClass)
  }
}

class RxImplConfigGen @Inject() (_injector: Injector) extends Rx {
  def cron(specification: String): IObservable[RxCronEvent] = {
    cronSpecifications += specification
    new RxConfigGenStream[RxCronEvent](this.injector)
  }

  def injector = _injector

  def uploads: IObservable[RxUploadEvent] = {
    new RxImplConfigGen.RxConfigGenStream[RxUploadEvent](this.injector)
  }

  def taskqueue[T <: Serializable](queueName: String): IObserver[RxTask[T]] = {
    taskQueues += queueName
    new IObserver[RxTask[T]] {
      def onCompleted(): Unit = {
        throw new UnsupportedOperationException
      }

      def onError(e: Exception): Unit = {
        throw new UnsupportedOperationException
      }

      def onNext(value: RxTask[T]): Unit = {
        throw new UnsupportedOperationException
      }
    }
  }

  def tasks[T <: Serializable](uploads: String, payloadClass: Class[T]): IObservable[RxTask[T]] = {
    new RxImplConfigGen.RxConfigGenStream[RxTask[T]](this.injector)
  }

  def tasks[T <: Serializable](queueName: String, typeToken: TypeToken[T]): IObservable[RxTask[T]] = {
    new RxImplConfigGen.RxConfigGenStream[RxTask[T]](this.injector)
  }

  def updates: IObservable[RxVersionUpdateEvent] = {
    new RxImplConfigGen.RxConfigGenStream[RxVersionUpdateEvent](this.injector)
  }

  def initialized: IObservable[RxInitializationEvent] = {
    new RxImplConfigGen.RxConfigGenStream[RxInitializationEvent](injector)
  }

  def generateConfigs(): Unit = {
    System.out.println("----- cron.xml -----")
    System.out.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<cronentries>")
    for (cronSpecification <- cronSpecifications) {
      System.out.println(s"    <cron>\n        <url>${RxImpl.getCronUrl(cronSpecification)}</url>\n        <schedule>$cronSpecification</schedule>\n    </cron>")
    }
    System.out.println("</cronentries>\n")
    System.out.println("----- queue.xml -----")
    System.out.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<queue-entries>")
    for (queue <- taskQueues) {
      System.out.println(s"    <queue>\n        <name>$queue</name>\n    </queue>")
    }
    System.out.println("</queue-entries>\n")
  }

  private var cronSpecifications: Set[String] = new HashSet[String]
  private var taskQueues: Set[String] = new HashSet[String]
}